crates/exex/exex/src/notifications.rs

```diff
@@ -13,101 +13,179 @@ use std::{
-    /// no remaining messages in the channel's buffer. This indicates that no
-    /// further values can ever be received from this `Receiver`. The channel is
-    /// closed when all senders have been dropped, or when [`Receiver::close`] is called.
+    /// Sets [`ExExNotifications`] to a stream of [`ExExNotification`]s without a head.
     ///
-    /// # Cancel safety
+    /// It's a no-op if the stream has already been configured without a head.
     ///
-    /// This method is cancel safe. If `recv` is used as the event in a
-    /// [`tokio::select!`] statement and some other branch
-    /// completes first, it is guaranteed that no messages were received on this
-    /// channel.
-    ///
-    /// For full documentation, see [`Receiver::recv`].
-    #[deprecated(note = "use `ExExNotifications::next` and its `Stream` implementation instead")]
-    pub async fn recv(&mut self) -> Option<ExExNotification> {
-        self.notifications.recv().await
+    /// See the documentation of [`ExExNotificationsWithoutHead`] for more details.
+    pub fn set_without_head(&mut self) {
+        let current = std::mem::replace(&mut self.inner, ExExNotificationsInner::Invalid);
+        self.inner = ExExNotificationsInner::WithoutHead(match current {
+            ExExNotificationsInner::Invalid => unreachable!(),
+            ExExNotificationsInner::WithoutHead(notifications) => notifications,
+            ExExNotificationsInner::WithHead(notifications) => ExExNotificationsWithoutHead::new(
+                notifications.node_head,
+                notifications.provider,
+                notifications.executor,
+                notifications.notifications,
+                notifications.wal_handle,
+            ),
+        });
     }
 
-    /// Polls to receive the next message on this channel.
-    ///
-    /// This method returns:
+    /// Returns a new [`ExExNotifications`] without a head.
     ///
-    ///  * `Poll::Pending` if no messages are available but the channel is not closed, or if a
-    ///    spurious failure happens.
-    ///  * `Poll::Ready(Some(message))` if a message is available.
-    ///  * `Poll::Ready(None)` if the channel has been closed and all messages sent before it was
-    ///    closed have been received.
+    /// See the documentation of [`ExExNotificationsWithoutHead`] for more details.
+    pub fn without_head(mut self) -> Self {
+        self.inner = ExExNotificationsInner::WithoutHead(match self.inner {
+            ExExNotificationsInner::Invalid => unreachable!(),
+            ExExNotificationsInner::WithoutHead(notifications) => notifications,
+            ExExNotificationsInner::WithHead(notifications) => ExExNotificationsWithoutHead::new(
+                notifications.node_head,
+                notifications.provider,
+                notifications.executor,
+                notifications.notifications,
+                notifications.wal_handle,
+            ),
+        });
```

Comment:
isn't this just self.set_without_head()?

Reply:
yep, fixed